### ISSUE TITLE
[Feature Request] Delete functionality? for soft delete entries

### ISSUE THREAD

**Issue by mpixelz** [Laravel-Backpack/CRUD#2282]
Hi,
In the case where the model has soft delete enabled, once we delete the entry it puts it to the deleted state.. but there is no way to either restore it? or permanently delete it?

i can only see delete button (again) even on the trashed entries which gives out error as its not actually forcing delete but soft deleting again.

am i the only one have this issue? or is it really not covered in default backpack crud?

**Comment by pxpm** [comment]
Hey @mpixelz 

I think atm is not provided. Anyway, I think is something we can provide with ease.

From my understanding of the problem I think we just need to update Delete trait, specificaly the delete() function.

Instead of:
```
public function delete($id)
    {
        return (string) $this->model->findOrFail($id)->delete();
    }
```

We could use something like this: 
```
public function delete($id)
    {
        if(in_array('Illuminate\Database\Eloquent\SoftDeletes', class_uses($this->model))) {
            $entry = $this->model->withTrashed()->findOrFail($id);
        }else{
            $entry = $this->model->findOrFail($id);
        }  
        
        if($entry->trashed()) {
            return (string) $entry->forceDelete();
        }else{
            return (string) $entry->delete();
        }      
    }
```

You could give it a shot in your project to test it out? 

Best,
Pedro

**Comment by mpixelz** [comment]
i have a solutions working for me.. as i have given option to restore the entry as well.. but i was thinking this is pretty basic and it would have been much better if it was done by backpack.. delete entry, if softdelete then option to either restore or permanently delete the entry.

**Comment by pxpm** [comment]
Hey @mpixelz 

Nice you already have it working. But as you said, maybe it's something we can take into account on core and probably would save other developers some time. :+1: 

Would you mind sharing your implementation for delete ? Is it too different from mine ? Have you tested mine in delete button ? 

Best,
Pedro

**Comment by tabacitu** [comment]
Hi guys,

I agree with you, I think it's something we could easily provide in 4.0 or 4.1: I see it as a "SoftDelete" operation, that you could use instead of the default "Delete" operation. 

# Solution 1 - multiple buttons

The operation would have:
- the current delete button (would do soft delete)
- a new "Hard Delete" button if the entry is deleted
- a new "Recover" button if the entry is deleted
- a new "Trashed" filter, to allow the admin to see the trashed items

All buttons should still have confirmation dialogs, of course.

# Solution 2 - dropdown button

Alternatively, the buttons above could all be merged inside one dropdown button, that shows:

[Delete ▼ ] -> soft deletes
- Move to Trash -> same thing as above, greyed out if it's already trashed
- Recover -> greyed out if it's not trashed
- Hard Delete

All buttons should still have confirmation dialogs, of course.

----

I think that's a complete implementation, from the admin's perspective, right?

**Solution 1** (multiple buttons) would make it impossible for the admin to hard delete an entry by mistake, but it would also make it a little more difficult to hard delete an entry if you know that's what you want (move to trash, then delete from trash).

**Solution 2** (dropdown button) would make it easy to "hard delete" an entry from the same screen, but it adds the chance that an admin mistakenly hard deletes an entry.

What do you think? 

PS. If we add this operation we also need to add a "BulkSoftDelete" operation.

**Comment by tabacitu** [comment]
Related to #1827

**Comment by DelightedD0D** [comment]
Personally, I'd prefer the cleaner UX of having the drop down with all my options available.

**Comment by mpixelz** [comment]
> Hey @mpixelz
> 
> I think atm is not provided. Anyway, I think is something we can provide with ease.
> 
> From my understanding of the problem I think we just need to update Delete trait, specificaly the delete() function.
> 
> Instead of:
> 
> ```
> public function delete($id)
>     {
>         return (string) $this->model->findOrFail($id)->delete();
>     }
> ```
> 
> We could use something like this:
> 
> ```
> public function delete($id)
>     {
>         if(in_array('Illuminate\Database\Eloquent\SoftDeletes', class_uses($this->model))) {
>             $entry = $this->model->withTrashed()->findOrFail($id);
>         }else{
>             $entry = $this->model->findOrFail($id);
>         }  
>         
>         if($entry->trashed()) {
>             return (string) $entry->forceDelete();
>         }else{
>             return (string) $entry->delete();
>         }      
>     }
> ```
> 
> You could give it a shot in your project to test it out?
> 
> Best,
> Pedro

not working

**Comment by mpixelz** [comment]
i would prefer solution 1.
reason: 
it would allow us to restrict direct hard delete. once the user deletes the item, they will still have the option to restore/force delete from trashed items.. and if we have restrictions on trash filter we can prevent hard delete with some more checks.
i hope im making sense here.

**Comment by pxpm** [comment]
@mpixelz could you let me know what error you get, why is it not working for you ? 

I would go for solution solution 1.5 ! I would not create the trashed item filter (That's something dev could do). I would had a filter in the button, if item is trashed (deleted_at is not null) the buttion would become `HARD DELETE` maybe with a red border. If not the default button is shown. 
I would not over-engineer here, change the function to delete trashed and non-trashed and give the button some visual indication. 

Best,
Pedro

**Comment by tabacitu** [comment]
@pxpm I agree with the sentiment to not overengineer this 😄 But with no Trashed filter, the hard deleted items would never be shown. By default, Backpack doesn't use ```withTrashed()``` when querying items...

**Comment by pxpm** [comment]
@tabacitu ahh sure, I misunderstood things in first place, I'm sorry. 

Sure the filter can be added to core filters. The thing is we can have a check in delete button that if the `$entry` is trashed we apply a red border and the text changes to `Force Delete` or `Hard Delete`  Without the dropdowns or wtv. I think I know why you see the dropdown being useful, like: 

**I know i want to hard delete something, and have to delete first, filter deleted and delete again** 

One solution, in case model uses SoftDeletes could be to have a checkbox in the delete confirmation dialog `[x] - Permanently delete this item?`  

Let me know your thoughts about it.

Best,
Pedro
